### DISCUSSION TITLE
Isar instance creation failure

### DISCUSSION THREAD

**Discussion by Biplab-Dutta** [isar/isar#688] [Help] [answered]
I have a bootstrap method that is called inside my `main()` function.
```dart
typedef _BootstrapBuilder = Widget Function(Dio dio);

void bootstrap(_BootstrapBuilder builder) {
  WidgetsFlutterBinding.ensureInitialized();

  Bloc.observer = AppBlocObserver();

  FlutterError.onError = (details) {
    log(details.exceptionAsString(), stackTrace: details.stack);
  };

  runZonedGuarded(
    () async {
      await IsarDatabase().init();   // Initializes isar database
      final dio = Dio()
        ..options = BaseOptions(
          baseUrl: 'api.github.com',
          headers: {'Accept': 'application/vnd.github.html+json'},
          responseType: ResponseType.plain,
          validateStatus: (status) {
            return status != null && status >= 200 && status < 400;
          },
        );
      runApp(builder(dio));
    },
    (error, stackTrace) => log(error.toString(), stackTrace: stackTrace),
  );
}
```
As you can see, I have called `init()` method that initializes my isar database.
```dart
class IsarDatabase {
  late Isar _instance;

  Isar get instance => _instance;

  /// Initializes the isar database.
  ///
  /// This method needs to be called before accessing any isar-specific APIs.
  Future<void> init() async {
    if (Isar.instanceNames.isNotEmpty) return;
    _instance = await Isar.open(
      [GithubHeaderSchema, DetailDTOSchema],
      inspector: false,
    );
  }
}
```
When I try to invoke a method that uses Isar API, I get this error in my console.
```
[log] LateInitializationError: Field '_instance@34001629' has not been initialized.
```
which points to Isar object.
Since, I am calling my bootstrap function in the main method, I was expecting my isar database to be initialized when I begin invoking methods in my data layers. But as you can see, that's not the case.

Can someone help me?
Thanks!

**Comment by sarbagyastha** [comment, marked as answer]
@Biplab-Dutta Is `IsarDatabase` a singleton ? Because you're calling `init` on a new instance of `IsarDatabase` inside bootstrap method.

**Reply by Biplab-Dutta** [reply]
No, `IsarDatabase` isn't a singleton.. it is a regular class.. i have shared the code for `IsarDatabase` class above..

Should I create a singleton?

**Reply by sarbagyastha** [reply]
How are you accessing `instance` getter from `IsarDatabase` in other part of your code ?

**Reply by Biplab-Dutta** [reply]
By passing IsarDatabase as a dependency in constructor..

And now I think I understood my issue

I suddenly remember the way I have written my constructors..

```dart
class DetailLocalService { 
   DetailLocalService({ 
     IsarDatabase? isarDb, 
   }) : _isarDb = isarDb ?? IsarDatabase(); 
  
   final IsarDatabase _isarDb; 
  
   Isar get _isar => _isarDb.instance;
```

So, I was doing something like this..
And now I think the object used to call `init()` inside the bootstrap and the object i am using here are different.. 😅

Maybe a singleton should fix this?

**Reply by sarbagyastha** [reply]
Yes, that's the issue. You're `init`-ing on one instance of `IsarDatabase` and trying to access it from different instance, which is not possible unless it's a singleton.

**Reply by Biplab-Dutta** [reply]
Thank you again.. 😅
